### PR TITLE
Code cleanup&organization, Evolve logic fixed once again.

### PR DESCRIPTION
**Pull request by ChemiHatake**
## Description:
Organized the NecroBot.CLI code entirely, 
(will work on NecroBot.Logic later)

Fixed the evolve logic, Please, before you close my pr because you think this logic is wrong,
READ THE WIKI

[Link to the wiki](https://github.com/NECROBOTIO/NecroBot/wiki/Config)

and if you are too lazy to do that, 
leave the work for someone else that has an actual reason to close or merge this PR,
thanks.

## Fixes:
#4528
#4440
#4399
#4258

(will work faster now!

### CODE REVIEW

**Comment by ChemiHatake** [comment]
Ill resolve the conflicts

**Comment by vcafezinho** [comment]
@ChemiHatake Thank you very much for fix it again. Even someone messing you hard work you still motivated redoing something that was already fixed.

**Comment by GrimmGringo** [comment]
No, you are not understanding what the configuration setting is meant to do:
UseLuckyEggsMinPokemonAmount (Value)

Value used when evolving to determine the number of evolutions needed to trigger a Lucky Egg use.

Let me rephrase it for you, if you've met the conditions for evolving X number of pokemon, THEN you can check if the X>=UseLuckyEggsMinPokemonAmount. Meaning, if the minimum amount of pokemon in order to use lucky eggs (UseLuckyEggsMinPokemonAmount) is met, then you will use a lucky egg. The setting is NOT saying to evolve and use an egg if you have that many pokemon ready to evolve, and evolve them.

**Comment by ChemiHatake** [comment]
@GrimmGringo DID YOU READ THE WIKI?

**Comment by ChemiHatake** [comment]
I'm totally done, I WROTE SPECIFICALLY "READ THE WIKI"
but he didn't.
you know what?
If you don't want to read it, Ill quote it

'UseLuckyEggsWhileEvolving (Value)

When UseLuckyEggsWhileEvolving is set to true, NecroBot will hold-off on evolving Pokemon until there are enough evolutions to meet the value in useLuckyEggsMinPokemonAmount and then proceed to use a Lucky Egg before evolving them. Before triggering the evolve step, the bot will use a lucky egg if one has not been used in the past 30 minutes.

UseLuckyEggsMinPokemonAmount (Value)

Value used when evolving to determine the number of evolutions needed to trigger a Lucky Egg use.'

**Comment by vcafezinho** [comment]
@GrimmGringo Grimm the problem is that now this is not working. Instead of keep the pokemons > use luck egg > evolve all, what is doing is evolving and not holding to evolve them all. No matter how we configure this.
image

What we what is: SET UP THE POKEMONS TO KEEP (PokemonsToEvolve) > USE LUCK EGG WHEN   UseLuckyEggsMinPokemonAmount" AND EVOLVE ALL TOGETHER.

**Comment by weedorbeat** [comment]
Merge this code, please! Also fixed #3811
